### ISSUE TITLE
NullPointerException in FileSystemWriter

### ISSUE THREAD

**Issue by rvanheest** [LibraryOfCongress/bagit-java#35]
I get a NullPointer from FileSystemWriter, line 232, who's call originates from line 222 and PreBagImpl, line 134, when I want to create a bag with empty/no payload.

```scala
val inputDir = multiDepositDir(settings, datasetID)
val inputDirExists = inputDir.exists
val outputBagDir = // the output directory for the bag

val bagFactory = new BagFactory
val preBag = bagFactory.createPreBag(outputBagDir)
val bag = bagFactory.createBag(outputBagDir)

// no bag.addFilesToPayload(...), since I want an empty payload
bag.makeComplete

val fsw = new FileSystemWriter(bagFactory)
fsw.setTagFilesOnly(true)
fsw.write(bag, outputBagDir)

preBag.makeBagInPlace(Version.V0_97, false)
```

The problem originates in the last line while calling `preBag.makeBagInPlace(Version.V0_97, false)` and finally reaches the point of failure in `FileSystemWriter` at line 232 in `dir.listFiles()`, which evaluates to `null`. According to the documentation on this expression, this should only happen if `dir` does not exist or if an `IOException` occurs. I assume the former one is the case here, which would easily be dealt with by checking for it's existence before calling `removeExtraFiles` with this directory (call is in line 222 of `FileSystemWriter`).

Is this observation correct and can you fix this?

**Comment by johnscancella** [comment]
@rvanheest some questions for you:
* which version of bagit-java are you using?
* `val outputBagDir = ` doesn't seem to ever be set, is there missing code?
* The code syntax looks like groovy, is that correct?
* Have you taken a look at our BETA release? That is the future of this library, and might be easier to deal with then the current release. You can download it from our [snapshots repo](https://oss.sonatype.org/content/repositories/snapshots/gov/loc/bagit/) or on our releases page [bagit-5.0-BETA](https://github.com/LibraryOfCongress/bagit-java/releases/tag/v5.0.0-BETA)

**Comment by rvanheest** [comment]
* We're currently on version 4.9. I see that you released a version 4.11 a couple of days ago though, so I'll have a look at that tomorrow.
* `val outputBagDir` is set, but you can add any valid file path
* I indeed don't use Java. We use Scala in our projects instead. Probably should mention that next time :smiley:
* Regarding the 5.0-BETA release: I'm curious to know what the differences are with the 4.x versions. I don't really see any documentation about this. Did the API stay the same or does it affect the use of the library as well?

Another general remark about this library: one of the things I really miss is good documentation. I was happy to have #18 as a source of 'how-to-use-this-library' when I started using your library some time ago, but I couldn't find any other 'official' documentation on this GitHub repo. It would really help if you as an author of this library gave more priority to the documentation. After all, you know how the library is intended to be used...

In short, I'll check out version 4.11 tomorrow and I'll come back to this issue afterwards.

**Comment by johnscancella** [comment]
* there isn't much difference between 4.9 and 4.11 except for some bug fixes.
* I haven't done any Scala programming so I will defer to you on if outputBagDir is set. 
* The difference between the 5.* and 4.* is that the 5.* is a complete rewrite of the library. Knowing about #18 I specifically added a bunch of examples in the new [README](https://github.com/LibraryOfCongress/bagit-java/blob/rewrite/README.md#examples). I understand the frustration on documentation(one of the reasons I decided to completely rewrite it), which is why I also have added a lot to the javadocs, so if you use an IDE you should get much more info. Hopefully with the new examples and the expansion of the javadocs it should be much easier to get started.

The 5.0-BETA is ready for testing, but it doesn't seem many people are using it yet. In the coming days I will most likely be using it in a new [Bagger](https://github.com/LibraryOfCongress/bagger) application, but if you want to try it and report any problems you have (or even better submit a pull request) I would be most grateful.

**Comment by rvanheest** [comment]
I just switched from 4.9 to 4.11 and as expected this solved the nullpointer bug. So that's great!

However, what I notice now is that the `data` folder is not generated at all, presumably due to the bag being empty/having no payload. Is that the expected behavior? At first glance I would expect an empty `data` folder to be there, but version 4.11 doesn't do that. In the [BagIt documentation](https://tools.ietf.org/html/draft-kunze-bagit-13#section-2.1.2) however, I read that this `data` folder MUST exist. Am I doing something wrong here?

**Comment by rvanheest** [comment]
Regarding the latter, I only now notice that also **NO** `manifest-md5.txt` file is created for a bag with no payload. Again, in the [documentation](https://tools.ietf.org/html/draft-kunze-bagit-13#section-2.1.3) I read "*Every bag MUST contain one payload manifest file, ...*".

**Comment by johnscancella** [comment]
@rvanheest You are correct, most likely this is a bug. However, could you please open a new issue since the null pointer issue is now resolved?

Also, I would highly encourage you to try out the 5.0-BETA since it would be much easier to do what you are asking:
```java
Version version = new Version(0, 97);
Bag bag = new Bag(version);
bag.getPayLoadManifests().add(new Manifest(StandardSupportedAlgorithms.MD5));
File outputDir = new File("LocationOfEmptyBagYouWant");
try {
  BagWriter.write(bag, outputDir);
} catch (IOException e) {
  e.printStackTrace();
}
```

**Comment by rvanheest** [comment]
Thanks for your help so far, mate! I made a new issue for that new bug, see #36.

I'll discuss the use of v5.0 in our company next week. Code looks great though!

**Comment by johnscancella** [comment]
No problem, glad I could be of help. Yeah I took it over after the previous developer left, and thought I could do it better. So I just rewrote all of it, and added lots of test to make sure it is compliant with the spec. Just create a new issue if you find any problems with the BETA code. After a couple of weeks I will probably put it out as a RC
